### PR TITLE
Transformations: groupBy transformation now preserves data order

### LINKED ISSUE #43545: groupBy transformer changed the order of numeric keys
**What happened**:

https://github.com/grafana/grafana/blob/c3eb1ffe850be4f3ee4093987160fbb29e196e2f/packages/grafana-data/src/transformations/transformers/groupBy.ts#L66

In groupBy transform, the values with same key will be clac.
But when the key was numeric string, Object.keys will change the order of numeric keys.

Can see [the-order-of-js-object-keys](https://dev.to/frehner/the-order-of-js-object-keys-458d)

**What you expected to happen**:

Consider use Map to store key-value paris to keep order of values.

**How to reproduce it (as minimally and precisely as possible)**:

image

data 
```csv
key,value
500,1000
404,2000
200,5000
204,10000
```

image

transform config
```json
  "transformations": [
    {
      "id": "groupBy",
      "options": {
        "fields": {
          "key": {
            "aggregations": [],
            "operation": "groupby"
          },
          "value": {
            "aggregations": [
              "first"
            ],
            "operation": "aggregate"
          }
        }
      }
    }
  ],
```

**Anything else we need to know?**:

**Environment**:
- Grafana version: latest 8.3
- Data source type & version: any
- OS Grafana is installed on: play.grafana.org
- User OS & Browser: macos chrome

### PR DESCRIPTION
**Pull request by MeetzhDing**
**What this PR does / why we need it**:

groupBy transformer changed the order of numeric keys

See: https://github.com/grafana/grafana/issues/43545

**Which issue(s) this PR fixes**:

Fixes https://github.com/grafana/grafana/issues/43545

**Special notes for your reviewer**:

My local environment get this https://github.com/facebook/jest/issues/9021 when running `yarn jest`, maybe you need to run the test yourself

### CODE REVIEW

File: packages/grafana-data/src/transformations/transformers/groupBy.ts [hunk 1]
```diff
@@ -89,15 +89,15 @@ export const groupByTransformer: DataTransformerInfo<GroupByTransformerOptions>
           }
 
           const fields: Field[] = [];
-          const groupKeys = Object.keys(valuesByGroupKey);
+          const groupKeys = Array.from(valuesByGroupKey.keys());
```
Comment (ashharrison90) [inline packages/grafana-data/src/transformations/transformers/groupBy.ts:92]: remove this line entirely

File: packages/grafana-data/src/transformations/transformers/groupBy.ts [hunk 2]
```diff
@@ -89,15 +89,15 @@ export const groupByTransformer: DataTransformerInfo<GroupByTransformerOptions>
...
 
           for (const field of groupByFields) {
             const values = new ArrayVector();
             const fieldName = getFieldDisplayName(field);
 
             for (let key of groupKeys) {
```
Comment (ashharrison90) [inline packages/grafana-data/src/transformations/transformers/groupBy.ts:98]: loop over `valuesByGroupKeys` here, e.g. `valuesByGroupKeys.forEach((value) => {`

File: packages/grafana-data/src/transformations/transformers/groupBy.ts [hunk 3]
```diff
@@ -89,15 +89,15 @@ export const groupByTransformer: DataTransformerInfo<GroupByTransformerOptions>
...
-              const valuesByField = valuesByGroupKey[key];
-              values.add(valuesByField[fieldName].values.get(0));
+              const valuesByField = valuesByGroupKey.get(key);
+              values.add(valuesByField![fieldName].values.get(0));
```
Comment (ashharrison90) [inline packages/grafana-data/src/transformations/transformers/groupBy.ts:100]:
```suggestion
              values.add(value[fieldName].values.get(0));
```

File: packages/grafana-data/src/transformations/transformers/groupBy.ts [hunk 4]
```diff
@@ -121,7 +121,7 @@ export const groupByTransformer: DataTransformerInfo<GroupByTransformerOptions>
             const valuesByAggregation: Record<string, any[]> = {};
 
             for (const groupKey of groupKeys) {
```
Comment (ashharrison90) [inline packages/grafana-data/src/transformations/transformers/groupBy.ts:123]: loop over `valuesByGroupKeys` here, e.g. `valuesByGroupKeys.forEach((value) => {`

File: packages/grafana-data/src/transformations/transformers/groupBy.ts [hunk 5]
```diff
@@ -121,7 +121,7 @@ export const groupByTransformer: DataTransformerInfo<GroupByTransformerOptions>
...
-              const fieldWithValuesForGroup = valuesByGroupKey[groupKey][fieldName];
+              const fieldWithValuesForGroup = valuesByGroupKey.get(groupKey)![fieldName];
```
Comment (ashharrison90) [inline packages/grafana-data/src/transformations/transformers/groupBy.ts:124]:
```suggestion
              const fieldWithValuesForGroup = value[fieldName];
```

**Review by ashharrison90 — COMMENTED** [review]
looks good - i think with a minor tweak we can avoid having to put `!` everywhere to please typescript.

the `groupKeys.length` on 153 would then become `valuesByGroupKeys.size` instead i think

wdyt? 😄

**Comment by MeetzhDing** [comment]
Looks great!

**Review by ashharrison90 — APPROVED** [review]
awesome contribution @MeetzhDing 🙌

**Pull request merged**